### ISSUE TITLE
xdg-desktop-portal-wlr capture sources (screen/window) can only ever be one and the same source

### ISSUE THREAD

**Issue by Omar007** [obsproject/obs-studio#4815]
### Operating System Info

Other

### Other OS

Arch Linux

### OBS Studio Version

27.0.0

### OBS Studio Log URL

N/A

### Expected Behavior

Each individual (new) Pipewire capture element to be isolated/unique in its configuration; each new Screen Capture can be a different screen and each new Window Capture can be a different window without affecting existing elements.

### Current Behavior

After adding a new Pipewire capture element and selecting a different screen or window from the previously existing Pipewire capture elements, the old/existing ones are also changed to the new selection.

### Steps to Reproduce

1. Add a new Screen/Window Capture (Pipewire) element
2. Select Monitor/Screen 1
3. Add a new Screen/Window Capture (Pipewire) element
4. Select Monitor/Screen 2
5. See the first capture element also change to Monitor/Screen 2

### Anything else we should know?

It's not just limited only to the one previously added element, you can continue adding for as long as you want; all existing ones will change to the new selection.  
When you remove the most recently added capture element, the previous ones will switch back to the screen you selected before it. You can also keep doing this for as long as you want.  
So if you have 3 screens (A, B, C) and add a unique/new capture element for each, first you have 1 with A, then 1+2 with B, then 1+2+3 with C at the end. Removing in the order 3, 2, 1 results in the reverse. Removing in the order 1, 2, 3 only ever keeps C showing, removing in the order 3, 1, 2 (or 1, 3, 2) makes it end with B (the selection for 2) never showing A again.

Aside from the above behaviour, it also looks like the Window Capture element is (doing) the same as the Screen Capture (both capture screen/output) but that looks to me to be a different issue, most likely not even an OBS issue.  
Probably a problem with XDG Desktop Portal wlr and/or the selection tool being used (slurp) as both elements result in the same selection command execution (`slurp -f %o -or`) and thus returned data; the output/screen name).

Running Sway (wlroots based Wayland).

**Comment by kkartaltepe** [comment]
Provide a log from obs.

**Comment by Omar007** [comment]
'A' log or one that contains something specific/with a log settings or something? Because they show absolutely nothing in relation to the described behaviour at all. For the removal even less.

```
19:20:40.461: User added scene 'test'
19:20:40.461: User switched to scene 'test'
19:20:45.167: adding 64 milliseconds of audio buffering, total audio buffering is now 106 milliseconds (source: Mic/Aux)
19:20:45.167: 
19:20:48.292: PipeWire initialized (sender name: 1_397)
19:20:48.294: User added source 'Screen Capture (PipeWire)' (pipewire-desktop-capture-source) to scene 'test'
19:20:48.322: [pipewire] available cursor modes:
19:20:48.322: [pipewire]     - Always visible
19:20:48.322: [pipewire]     - Hidden
19:20:48.326: [pipewire] screencast session created
19:20:49.395: [pipewire] asking for desktop…
19:20:49.402: [pipewire] desktop selected, setting up screencast
19:20:49.408: [pipewire] created stream 0x555b5b448f30
19:20:49.408: [pipewire] playing stream…
19:20:57.588: PipeWire initialized (sender name: 1_397)
19:20:57.590: User added source 'Screen Capture (PipeWire) 2' (pipewire-desktop-capture-source) to scene 'test'
19:20:57.607: [pipewire] available cursor modes:
19:20:57.607: [pipewire]     - Always visible
19:20:57.607: [pipewire]     - Hidden
19:20:57.610: [pipewire] screencast session created
19:21:02.235: [pipewire] asking for desktop…
19:21:02.241: [pipewire] desktop selected, setting up screencast
19:21:02.288: [pipewire] created stream 0x555b5b48c700
19:21:02.288: [pipewire] playing stream…
19:21:13.647: User Removed source 'Screen Capture (PipeWire) 2' (pipewire-desktop-capture-source) from scene 'test'
19:21:21.676: User Removed source 'Screen Capture (PipeWire)' (pipewire-desktop-capture-source) from scene 'test'
19:21:25.038: User switched to scene 'Break / AFK'
19:21:25.038: User Removed scene 'test'
```

**Comment by kkartaltepe** [comment]
Just like the template specifies, a full log from when you experienced the issue. Please do not trim the log and provide it as is unaltered.

**Comment by Omar007** [comment]
Template didn't specify and only talked about a URL which I don't have? I assume if the application crashes it allows you to upload something or w/e but it's not crashing or anything. Just pulled this from `~/.config/obs-studio/logs`  :man_shrugging: 

I'll attach the full thing when I get back.

**Comment by GeorgesStavracas** [comment]
It seems to me that it's an issue in wlroots. On OBS side, each capture source spawns one separate thread and PipeWire loop, and they basically don't interact with each other. This seems like the internal state of xdg-desktop-portal-wlr gets confused with two streams.

**Comment by Omar007** [comment]
That sounds like it could be plausible. I did not find anything about such a limitation but that doesn't eliminate the possibility it isn't an upstream bug.  
I'll see if I can create a similar situation using XDG Desktop Portal directly to see if I can reproduce the problem there :thinking: 

The side-note/additional thing I mentioned in the "Anything else" segment (about the Window Capture == Screen Capture) is indeed xdg-desktop-portal-wlr related. Looks like that isn't supported yet and there is an open item for it https://github.com/emersion/xdg-desktop-portal-wlr/issues/107

**Comment by GeorgesStavracas** [comment]
We cannot do anything on OBS side to work around this missing implementation, but it made me realize showing the window capture there is wrong and fixable. Incoming PR soon.

**Comment by GeorgesStavracas** [comment]
@Omar007 would you please test the pull request above, and let us know if the "Window Capture (PipeWire)" entry is still visible on your machine?

**Comment by Omar007** [comment]
@GeorgesStavracas Sure, I can probably do so later today :+1:

**Comment by Omar007** [comment]
Looks like that works and hides the unsupported capture mode.

Before
![screenshot_2021-06-14T17:15:57](https://user-images.githubusercontent.com/5380079/121932832-9e777b00-cd45-11eb-883a-d7fe930072e3.png)

After
![screenshot_2021-06-14T17:16:16](https://user-images.githubusercontent.com/5380079/121932850-a3d4c580-cd45-11eb-8b66-c80d893c9b9b.png)

Windows Capture (second to last item) is not shown in the "After" :+1: 

On the subject of OBS related improvements, as another unrelated side-note, I also noticed on start the screen for Screen Capture needed to be reselected every time OBS was started with one of those in a scene.


Putting the side-notes aside for a bit and going back to the main/OG issue, I haven't gotten around to verifying the main issue outside of OBS yet so I can't confirm or deny whether or not the suspicion is correct. I should be able to do so this week. If it ends up being replicable outside of OBS I'll close this as an upstream issue.

**Comment by kkartaltepe** [comment]
> On the subject of OBS related improvements, as another unrelated side-note, I also noticed on start the screen for Screen Capture needed to be reselected every time OBS was started with one of those in a scene.

That is a limitation of the xdg-desktop-portal protocol.

**Comment by Omar007** [comment]
Ah alright, fair enough. I thought I read it can be done without a chooser (requiring the user to supply it directly which I figured could be done as a property/setting on the OBS Scene Capture component) but maybe that was something specifically for xdg-desktop-portal-wlr :man_shrugging:

**Comment by Omar007** [comment]
Alright I was able to do a test of the OG problem using both OBS and Firefox. Using OBS to select 1 screen, then running Firefox with a screenshare using https://mozilla.github.io/webrtc-landing/gum_test.html and selecting another, it's showing the same behaviour crossing program borders so it's def. an xdg-desktop-portal-wlr issue.

Thanks for all the info (and the tangent improvement :grin:  ). Closing this for being an upstream problem  :+1:
